portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -0,0 +1,58 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.gson.Gson;
+
+
+
+@WebServlet("/login")
+public class LoginServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String[] loginStatus = new String[3];
```

Comment:
An array of strings for `loginStatus` is reasonably readable here, but less so once you start using it in the script later. The reader has to flip back to this file to understand the meaning of each index, for example, and you end up using string comparison to check for a true/false value of whether the user is logged in. I think a custom abstraction would make it much easier to understand the meaning of the data.

http://go/tott/494

Reply:
The link you sent here is restricted for me.  Could you give a little more information on custom abstraction? I tried a couple searches, but couldn't find anything related.

Reply:
Sorry about the link. By "custom abstraction" I just mean that instead of an array of strings, you can create your own type to represent login status, and use appropriate types and descriptive field names within that type to represent the status, link, and text. So, instead of treating it as a bag of strings which you compare to other strings, you can treat it as an entity with named properties. That makes it easier to read the code without having to remember the meaning and structure of each string within `loginStatus`.

Reply:
Oh, okay. Makes perfect sense, I'll work on it now.